syntax = "proto3";

option java_package = "com.grpc.grpc_server.services";
option java_outer_classname = "ProductoManagerServiceProto";

import "producto.proto";
import "tienda.proto";

service ProductoManagerService {
  rpc FindById(ProductoEnTienda) returns (ProductoEnTienda);
  rpc FindAll(ProductoManagerEmpty) returns (ProductosEnTienda);
  rpc FindAllByProductoId(Producto) returns (ProductosEnTienda);
  rpc FindAllByTiendaId(Tienda) returns (ProductosEnTienda);
  rpc FindAllByCustomFilter(CustomFilter) returns (ProductosEnTienda);
  rpc AssingProductoToTienda(AssignToTiendaMessage) returns (AssignResponse);
  rpc AddProducto(ProductoBase) returns (ProductoEnTienda);
  rpc ModifyProducto(ProductoBase) returns (ProductoEnTienda);
  rpc ModifyStock(ProductoEnTienda) returns (ProductoEnTienda);
}

message ProductoManagerEmpty {}

message ProductosEnTienda {
  repeated ProductoEnTienda productoEnTienda = 1;
}

message AssignResponse {
    string message = 1;
}

message AssignToTiendaMessage {
  ProductoBase productoBase = 1;
  Tienda tienda = 2;
}

message CustomFilter {
    string nombre = 1;
    string codigo = 2;
    string talle = 3;
    string color = 4;
    Tienda tienda = 5;
}

message ProductoEnTienda {
    int32 id = 1;
    Producto producto = 2;
    Tienda tienda = 3;
    int32 stock = 4;
    string talle = 5;
    string color = 6;
}

message ProductoBase {
    int32 id = 1;
    string nombre = 2;
    string codigo = 3;
    string foto = 4;
    string talle = 5;
    string color = 6;
}